Atlas group update

yhanw0719 opened the pull request:
major changes:
1. latest atlas docker is "[ghcr.io/lbnl-science-it/atlas:v1.0.7](http://ghcr.io/lbnl-science-it/atlas:v1.0.7)
2. In run.py, create first skims at warm start atlas stage instead of warm start asim
3. In atlas preprocessor, add accessibility calculation step and add atlas_beamac in settings.yaml to turn this step on/off
4. solve potential atlas parallel errors by adding "run_atlas_auto", which will resubmit atlas run when it fails due to parallel errors
5. add "vehicleTypeId" column in atlas output for BEAM team to read as input
6. add "pilates/atlas/atlas_input/taz_to_tract_sfbay.csv", which is an input for atlas preprocessor to calculate accessibility

yhanw0719 committed:
Merge pull request #4 from LBNL-UCB-STI/atlas

pull latest commits from lbnl repo

yhanw0719 committed:
update function run_atlas_autoin run.py to deal with connection errors

atlas sometimes encounters "error reading from connection", which is related to parallel computing and can be fixed by a simple re-submission. Run_atlas_auto function will handle this re-submission when needed. I added "try except" statement to make sure it will not get broken when such errors happen.

yhanw0719 committed:
atlas postprocess vehicleTypeId for beam

yhanw0719 committed:
update atlas docker to 1.0.6

yhanw0719 committed:
update atlas postprocessor

since atlas docker is updated to v1.0.6, the "vintage_category" to "model_year" translation in atlas postprocessor can be removed now

yhanw0719 committed:
Merge branch 'atlas-yuhan-sfbay' into yhanw0719-patch-1

yhanw0719 committed:
Merge pull request #5 from LBNL-UCB-STI/yhanw0719-patch-1

keep track of LBNL-UCB-STI master + atlas postprocessor minor fix

yhanw0719 committed:
fix var name bug

atlas output "model_year" -> "modelyear"

yhanw0719 committed:
atlas: four updates

major changes:
1. create first skims at warm start atlas stage instead of warm start asim
2. add accessibility calculation in atlas preprocessor, and add atlas_beamac in settings.yaml to turn it on/off
3. solve potential parallel errors in atlas by adding an auto-resubmission when run fails due to parallel errors
4. add vehicleTypeId column in atlas output for BEAM team to read as input

yhanw0719 committed:
minor fix

correct a minor error in atlas preprocessor

yhanw0719 committed:
update atlas docker to v1.0.7

v1.0.7: atlas_beamac will be parsed to atlas docker, if 0, atlas will read from "accessibility_2015.RData", otherwise atlas preprocessor will calc accessibility itself.

yhanw0719 committed:
upload taz_to_tract_sfbay.csv

an atlas input needed for calculating accessibility in preprocessing stage (when atlas_beamac > 0)

yhanw0719 committed:
Merge branch 'main' of https://github.com/yhanw0719/PILATES

yhanw0719 committed:
Merge pull request #14 from yhanw0719/main

atlas group updates

zneedell commented on pilates/atlas/preprocessor.py:
```diff
@@ -103,3 +106,115 @@ def prepare_atlas_inputs(settings, year, warm_start=False):
+    ODmatrix = _get_time_ODmatrix(settings, path_list, measure_list, threshold)
+    
+    # assign values = 1 if time taken by public transit <= 30min; 0 if not
+    ODmatrix[ODmatrix<=30] = 1 
```
Might be marginally cleaner / faster to just overwrite this as a boolean array: `ODmatrix = ODmatrix <= 30`

yhanw0719 replied:
Ok! done

zneedell commented on pilates/atlas/preprocessor.py:
```diff
@@ -103,3 +106,115 @@ def prepare_atlas_inputs(settings, year, warm_start=False):
+    
+    # read OD matrix size (i.e., range of zone_id)
+    skims_dir = settings['asim_local_input_folder']
+    skims = omx.open_file(os.path.join(skims_dir, 'skims.omx'), mode = 'r') 
```
Not sure if this ends up being slow, but you could potentially get `zone_count` from the size of `ODmatrix` and/or return it from `_get_time_ODmatrix` so you don't have to open the omx file twice

yhanw0719 replied:
Thanks! done.

zneedell commented on pilates/atlas/preprocessor.py:
```diff
@@ -103,3 +106,115 @@ def prepare_atlas_inputs(settings, year, warm_start=False):
+    jobs['zone_id'] = jobs['block_id'].map(mapping)
+    
+    # count number of jobs for each block_id
+    jobs_vector = jobs.groupby('block_id').agg({'job_id':'size','zone_id':'max'}).rename(columns={'job_id':'access_sum'})
```
So just to confirm I'm thinking about it right--in `jobs.csv` do we have one row for every job? And we're taking the max of `zone_id` because sometimes a block can be associated with multiple zones but we want to just choose one, so we choose the highest zone id? (Seems a bit arbitrary but I'm sure it's fine if it's not a ton)

yhanw0719 replied:
Yes, we have one row for each job in jobs.csv and each job has its unique block_id. One block is associated with only one taz (according to sfbay geoid_to_zone.csv), so those zone_id should all be identical for the same block_id. I just used "max" to extract its value. "min" or "mean" can do the same job.

zneedell commented on pilates/atlas/preprocessor.py:
```diff
@@ -103,3 +106,115 @@ def prepare_atlas_inputs(settings, year, warm_start=False):
+    taz_to_tract = pd.read_csv('{}/taz_to_tract_{}.csv'.format(settings['atlas_host_input_folder'], settings['region']), index_col=0)
+    
+    # convert taz- to tract-level accessibility data
+    accessibility_tract = np.matmul(np.transpose(accessibility), np.array(taz_to_tract.values))
```
Is each tract always going to be associated with no more than one TAZ? If not I'd worry that we'd end up with some outputs that are the sum of two taz's accessibility counts, which I think might not be the intended behavior.

We could probably get around some of this (and also be a little more robust to indices being out of order) by using df.merge on `zone_id`.

yhanw0719 replied:
Yes, one tract can be associated with multiple tazs, and one taz can be associated with multiple tracts. 

This tract-taz conversion has been carefully handled using the "taz_to_tract_sfbay.csv" file we generated externally. It is basically a weight matrix (dim = 1454 taz x 1588 tracts). Weight for taz n and tract m is the land area of taz n inside tract m, divided by the total land area in tract m. By doing np.matmul, a 1x1454 taz accessibility vector is multiplied by this matrix, and generates a 1588x1 tract accessibility vector. That said, the accessibility of a tract (i.e., number of jobs accessible from this tract) is estimated as the weighted average of accessibility of its overlapping tazs. 

It is noteworthy that this conversion relies on "taz_to_tract_sfbay.csv", so beamac (in settings.yaml) should only be set to 1 if running sfbay taz, otherwise we recommend setting beamac = 0. I can add an error message and comment in run.py file to clarify this issue.

zneedell replied:
Ah, gotcha, thanks! I had assumed that tract-taz was a binary matrix but this makes a lot more sense and is a nice solution

zneedell reviewed (changes requested):
Looks great! A few small comments/questions, but none of them are blocking as long as it's working well in practice.

yhanw0719 committed:
simplification in atlas preprocessor

function not changed. Minor simplifications: 1) two-line commands simplified to one line "ODmatrix = ODmatrix<=30"; 2) get zone_count directly from ODmatrix size without opening skims.omx again

yhanw0719 commented:
Hi Zach, thanks for the careful review and comments. I have replied to each of them. Made two fixes and explained the other two.

yhanw0719 commented:
Hi Zach,
I just added a warning if atlas_beamac is set to 1 in non-sfbay cases.
Could you please merge that as well? Sorry for missing it in the last pull
request.
Best,
Yuhan

On Wed, Jun 8, 2022 at 1:02 PM Zach Needell ***@***.***>
wrote:

> Merged #15 <https://github.com/LBNL-UCB-STI/PILATES/pull/15> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LBNL-UCB-STI/PILATES/pull/15#event-6772751084>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO52CKPKQDFPXJ7NB7HZISTVOD36BANCNFSM5XSYZYWQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
